File: merge_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

void print(vector<int> &vec)
{
    for_each(vec.begin(), vec.end(), [](int i)
    {
        cout << i << " ";
    });
    cout << endl;
}

void Merge(vector<int> &vec, int left, int middle, int right)
{
    vector<int> left_vec(vec.begin() + left, vec.begin() + middle + 1);
    vector<int> right_vec(vec.begin() + middle + 1, vec.begin() + right + 1);
    int left_len = middle - left + 1;
    int right_len = right - middle;

    int i = 0, j = 0;
    int k = left;

    while (i < left_len && j < right_len)
    {
        if (left_vec[i] < right_vec[j])
        {
            vec[k++] = left_vec[i++];
        }
        else
        {
            vec[k++] = right_vec[j++];
        }
    }

    while (i < left_len)
    {
        vec[k++] = left_vec[i++];
    }
    while (j < right_len)
    {
        vec[k++] = right_vec[j++];
    }
}

void MergeSortImplement(vector<int> &vec, int left, int right)
{
    if (left < right)
    {
        int middle = (left + right) / 2;
        MergeSortImplement(vec, left, middle);
        MergeSortImplement(vec, middle + 1, right);
        Merge(vec, left, middle, right);
    }


}

void MergeSort(vector<int> &vec)
{
    MergeSortImplement(vec, 0, vec.size() - 1);
}


int main(void)
{
    srand((unsigned)time(NULL));
    vector<int> vec;
    for (int i = 0; i < 15; ++i)
    {
        vec.push_back(static_cast<int>(rand() % (34 - 2 + 1) + 2));

    }

    cout << "Before sort:          ";
    print(vec);

    vector<int> vec2 = vec;
    sort(vec2.begin(), vec2.end());
    cout << "Use generic function: ";
    print(vec2);

    MergeSort(vec);

    cout << "Use hand-by function: ";
    print(vec);

    return 0;
}

File: python_version/merge_sort.py
#!/usr/bin/env python3

def Merge(arr, left, middle, right):
    left_arr = arr[left: middle + 1] 
    right_arr = arr[middle + 1: right + 1]
    # print("left_arr: ", left_arr)
    # print("right_arr: ", right_arr)
    # print("arr: ", arr[left:right+1])
    left_size = middle - left + 1
    right_size = right - middle
    i, j = 0, 0
    k = left
    while i < left_size and j < right_size:
        if left_arr[i] < right_arr[j]:
            arr[k] = left_arr[i]
            i += 1
            k += 1
        else:
            arr[k] = right_arr[j]
            j += 1
            k += 1

    while i < left_size:
        arr[k] = left_arr[i]
        i += 1
        k += 1
    while j < right_size:
        arr[k] = right_arr[j]
        j += 1
        k += 1
    # print("after sort arr: ", arr[left:right+1])


def MergeSortImplement(arr, left, right):
    if left < right:
        middle = (left + right) // 2
        MergeSortImplement(arr, left, middle)
        MergeSortImplement(arr, middle + 1, right)
        Merge(arr, left, middle, right)


def MergeSort(arr):
    MergeSortImplement(arr, 0, len(arr) - 1)


arr = [2, 4, 1, 5, 12, 0, 23]
print(arr)
MergeSort(arr)
print(arr)
